### PR TITLE
#6 add response creation

### LINKED ISSUE #6: Add demo of slave-to-slave replication
Following from #3, the next steps are to add some basic replication demos.

For demoing "sync" functionality, the current idea is to do a proof-of-concept of "partial" replication to different sites. The simplest way to do this would be to simulate satellites by spinning up some CouchDB instances locally and assigning each satellite a local hostname which is mapped to a list of store codes. 

To simulate the slave-to-slave "syncing" process, can add/update a store which is shared across multiple sites. Can then send a document to the "satellite" replicator to trigger it to replicate this change up to the master. Assuming master-to-slave replication is setup on the "primary", the updated store data should then be synced back down to all the other sites on which that store is active.

Just rubber ducking here, so feedback welcome!

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #6. 

Adds event listener and updates satellite and primary selectors. A little bit of tidying up as I went along,  mostly just making variables a bit more explicit.

### CODE REVIEW

File: src/Listener.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  logResponse(doc: any) {
+    // tslint:disable-next-line:no-console
+    console.log(
+      `Response requisition with ID ${doc._id} created for request requistion with id ${doc.requestRequisitionId}\n`
+    );
+  }
+
+  listen(db) {
+    this.logListen();
+    const feed = db.follow({ since: "now" });
```
Comment (joshxg) [inline src/Listener.ts:31]: this is a nice feature theyv'e added. With a server dedicated to listening for changes, you could store when a server is taken down, or when it crashed and have a "catch up" mechanism

File: src/Listener.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+          this._nano.uuids(1).then((uuidDoc: any) => {
+            const { uuids } = uuidDoc;
+            const [uuid] = uuids;
+            responseRequisition._id = uuid;
+            responseRequisition.requestRequisitionId = requestRequisition._id;
+            responseRequisition.lines = requestRequisition.lines;
+            db.insert(responseRequisition).then((responseDoc: any) => {
+              if (responseDoc.ok) this.logResponse(responseRequisition);
+            });
+          });
```
Comment (joshxg) [inline src/Listener.ts:50]: this is wicked. Another idea is that if you had your graphql architecture as a federation https://www.apollographql.com/docs/apollo-server/federation/introduction/ - where you had a "requisition service", you could have the response requisition listener in that code base, maybe. Pretty cool imo!

File: scripts/sync.js [hunk 3]
```diff
@@ -46,39 +51,48 @@ const sync = async () => {
...
+        },
+      },
+      satelliteToPrimary: {
+        continuous: true,
+        selector: {
+          $and: [
+            { fromStoreId: { $in: activeStoreIds } },
+            {
+              $not: {
+                toStoreId: { $in: activeStoreIds },
```
Comment (joshxg) [inline scripts/sync.js:75]: for further iterations: I think you could basically have separate replicator documents for each 'module'  -  so you could have a 'requisition replication' document, which had all the logic for that. And, if there were ever customisations for a particular client (for example, I believe there were large changes made to requisitions and sync for ivory coast, who wanted requisitions editable in multiple places) , essentially making flexible sync code ... and also so easy to create a testing environment for !

File: scripts/sync.js [hunk 4]
```diff
@@ -15,19 +15,24 @@ const sync = async () => {
...
-    if (dbs.includes("_replicator")) await db.destroy("_replicator");
+    if (satelliteDatabases.includes("ocsupply"))
+      await satelliteDatabase.destroy("ocsupply");
+    await satelliteDatabase.create("ocsupply");
 
-    await db.create("ocsupply");
-    await db.create("_replicator");
+    if (satelliteDatabases.includes("_replicator"))
+      await satelliteDatabase.destroy("_replicator");
+    await satelliteDatabase.create("_replicator");
```
Comment (joshxg) [inline scripts/sync.js:31]: packages, packages, packages everywhere: https://github.com/jo/couchdb-ensure

Comment (wlthomson) [reply scripts/sync.js:31]: Nice... that guy has lots of couch repos!

Comment (wlthomson) [reply scripts/sync.js:31]: Nice... that guy has lots of couch repos!

File: scripts/sync.js (hunk 3 above)
Comment (wlthomson) [reply scripts/sync.js:75]: Yeah, I think the general pattern is to have separate docs for each "use-case" for the reasons you said... but we don't have enough use-cases!

File: src/Listener.ts (hunk 1 above)
Comment (wlthomson) [reply src/Listener.ts:31]: Yeah, exactly! Also resolves the problem of the listener being inaccessible if you wanted to have it as a micro-service or something like that.

**Pull request merged**